### ISSUE TITLE
Better type hinting

### ISSUE THREAD

**Issue by stancl** [php-plus/engine#5]
1. arrays of X

```php
public function addProducts(Product[] $products);
public function getProducts(): Product[];
```

2. `|` in typehints

```php
public function addProducts(Product[]|string[] $products):
```

3. arrays with X keys and Y values

```php
public function getProducts(): array<string, Product> // Product values, string keys
```

4. It would be nice to be able to say "`Collection` of `Product`s" (so a `Collection` with the first constructor argument being of type `Product[]`), but I have no idea how that should be notated.

**Comment by olivernybroe** [comment]
So this is basically adding union types and type hinted arrays.

I am not sure I am such a big fan of the key value syntax. Writing
```
Product[string]
```
Doesn't seem so clear that `string` is the key type. What about using the phpdoc way?
```
array<string, Product>
```

However I like your syntax when we don't use custom keys

**Comment by stancl** [comment]
> I am not sure I am such a big fan of the key value syntax

Yeah, It's a bit confusing to read, since the key comes after the value. But I think it's some convention.

> What about using the phpdoc way? `array<string, Product>`

Is this the phpdoc way? I haven't seen this yet.

Some people also do this:

```php
array[string]Product
```

**Comment by olivernybroe** [comment]
@stancl sorry, it's the Psalm way. https://psalm.dev/docs/annotating_code/type_syntax/array_types/

**Comment by stancl** [comment]
I see. Psalm uses `Foo[]` and `array<Foo, Bar>` so it's already a bit standardized. I agree we should go with that. Updated my original comment.
